src/morx.rs

```diff
@@ -0,0 +1,1825 @@
+                                    one_byte_values[usize::from(glyph - *first_glyph)] as u16,
+                                );
+                            } else {
+                                return None; //out of bounds
+                            }
+                        }
+                        UnitSize::TwoByte {
+                            lookup_values: two_byte_values,
+                        } => {
+                            if (glyph >= *first_glyph)
+                                && (glyph <= (*first_glyph + *glyph_count - 1))
+                            {
+                                return Some(two_byte_values[usize::from(glyph - *first_glyph)]);
+                            } else {
+                                return None; //out of bounds
+                            }
+                        }
+                        _ => return None, //Note: ignore 4-byte and 8-byte lookup values for now
+                    }
+                }
+                _ => return None, //Only Format10 is valid here.
+            }
+        }
+        _ => return None, //No more formats except the ones above
+    }
+}
+
+//------------------------------  Looking Up Class ------------------------------------------------------
+
+pub fn glyph_class(glyph: u16, class_table: &ClassLookupTable) -> u16 {
+    match lookup(glyph, class_table) {
+        None => {
+            return 1;
+        }
+        Some(val) => {
+            if val == 0xFFFF {
+                return 2;
+            } else {
+                return val;
+            }
+        }
+    }
+}
+
+//--------------------------------- Contextual Substitution ---------------------------------------------
+pub struct ContextualSubstitution<'a> {
+    glyphs: &'a mut Vec<RawGlyph<()>>,
+    next_state: u16,
+    mark: Option<(usize, u16)>, //record marked glyph and its position: (position, mark_glyph)
+}
+
+impl<'a> ContextualSubstitution<'a> {
+    pub fn new(glyphs: &'a mut Vec<RawGlyph<()>>) -> ContextualSubstitution<'a> {
+        ContextualSubstitution {
+            glyphs: glyphs,
+            next_state: 0,
+            mark: None,
+        }
+    }
+
+    pub fn process_glyphs(
+        &mut self,
+        contextual_subtable: &ContextualSubtable,
+    ) -> Result<(), ParseError> {
+        const SET_MARK: u16 = 0x8000;
+        const DONT_ADVANCE: u16 = 0x4000;
+        let mut old_glyph: u16;
+        let mut new_glyph: u16;
+
+        //loop through glyphs:
+        for i in 0..self.glyphs.len() {
+            let current_glyph: u16 = self.glyphs[i].glyph_index;
+            old_glyph = self.glyphs[i].glyph_index;
+            new_glyph = self.glyphs[i].glyph_index;
+
+            let mut class = glyph_class(current_glyph, &contextual_subtable.class_table);
+
+            'glyph: loop {
+                let index_to_entry_table = contextual_subtable.state_array.state_array
+                    [usize::from(self.next_state)][usize::from(class)];
```

Comment:
In places where we index into arrays with values provided by the font I think [`get`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get) should be used in order to avoid panics if the value happens to be out of bounds.